File: 0250-count-univalue-subtrees/0250-count-univalue-subtrees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    private int count;
    public int countUnivalSubtrees(TreeNode root) {
       countUntil(root);
       return count;
    }

    public boolean countUntil(TreeNode root) {
        if(root == null) 
            return true;

        boolean left = countUntil(root.left);
        boolean right = countUntil(root.right);

        if(left && right) {
            if(root.left != null && root.val != root.left.val) {
                return false;
            }

            if(root.right != null && root.val != root.right.val) {
                return false;
            }

            count++;
            return true;

        } else {
            return false;
        }    
    }
}

File: 0250-count-univalue-subtrees/0250-count-univalue-subtrees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    int countUnivalSubtrees(TreeNode* root) {
        int count[1];
        count[0] = 0;
        countUntil(root, count);
        return count[0];
    }

    bool countUntil(TreeNode* root, int count[]) {
        if (root == nullptr) {
            return true;
        }

        bool left = countUntil(root->left, count);
        bool right = countUntil(root->right, count);

        if (left & right) {
            if (root->left != nullptr && root->val != root->left->val) {
                return false;
            }
            if (root->right != nullptr && root->val != root->right->val) {
                return false;
            }

            count[0]++;
            return true;
        } else {
            return false;
        }
    }
};